File: realtime/realtime.py
import asyncio
import websockets
import json
import base64
import uuid
import threading
import sys
from typing import List, Dict, Any
import pyaudio

# Constants
WEBSOCKET_URI = "wss://api.openai.com/v1/assistants-streaming"
MODEL = "gpt-4o-realtime-preview-2024-10-01"
MODALITIES = ["text", "audio"]
VOICE = "alloy"
TRANSCRIPTION_MODEL = "whisper-1"
AUDIO_FORMAT = "pcm16"
TURN_DETECTION = {
    "type": "server_vad",
    "threshold": 0.5
}
SILENCE_DURATION_MS = 200

# Initialize PyAudio for audio playback
p = pyaudio.PyAudio()

def play_audio(audio_bytes: bytes):
    stream = p.open(format=pyaudio.paInt16,
                    channels=1,
                    rate=16000,
                    output=True)
    stream.write(audio_bytes)
    stream.stop_stream()
    stream.close()

class Client:
    def __init__(self):
        self.websocket = None
        self.session_id = None
        self.input_audio_buffer = bytearray()
        self.conversation: List[Dict[str, Any]] = []
        self.running = True
        self.lock = threading.Lock()

    async def connect(self):
        try:
            async with websockets.connect(WEBSOCKET_URI) as ws:
                self.websocket = ws
                await self.create_session()
                await asyncio.gather(
                    self.receive(),
                    self.handle_user_input()
                )
        except Exception as e:
            print(f"Error occurred: {e}")

    async def create_session(self):
        session_message = {
            "action": "create_session",
            "model": MODEL,
            "modalities": MODALITIES,
            "voice": VOICE,
            "input_audio_transcription": {
                "model": TRANSCRIPTION_MODEL,
                "format": AUDIO_FORMAT
            },
            "output_audio_format": AUDIO_FORMAT,
            "turn_detection": TURN_DETECTION,
            "silence_duration_ms": SILENCE_DURATION_MS
        }
        await self.websocket.send(json.dumps(session_message))
        print("Session creation message sent.")

    async def handle_user_input(self):
        loop = asyncio.get_event_loop()
        while self.running:
            user_input = await loop.run_in_executor(None, sys.stdin.readline)
            user_input = user_input.strip()
            if user_input.lower() == "/exit":
                await self.close_connection()
                break
            elif user_input.startswith("/truncate "):
                item_id = user_input.split(" ", 1)[1]
                await self.truncate_item(item_id)
            elif user_input.startswith("/delete "):
                item_id = user_input.split(" ", 1)[1]
                await self.delete_item(item_id)
            elif user_input.startswith("/audio "):
                # Assume audio input is provided as a file path
                file_path = user_input.split(" ", 1)[1]
                await self.handle_audio_input(file_path)
            else:
                await self.send_text_message(user_input)

    async def send_text_message(self, text: str):
        message = {
            "action": "send_conversation_item",
            "type": "text",
            "content": text,
            "id": str(uuid.uuid4())
        }
        self.conversation.append(message)
        await self.websocket.send(json.dumps(message))
        print(f"User said: {text}")

    async def handle_audio_input(self, file_path: str):
        try:
            with open(file_path, "rb") as f:
                audio_data = f.read()
                encoded_audio = base64.b64encode(audio_data).decode('utf-8')
                message = {
                    "action": "send_conversation_item",
                    "type": "audio",
                    "content": encoded_audio,
                    "id": str(uuid.uuid4())
                }
                self.conversation.append(message)
                await self.websocket.send(json.dumps(message))
                print("Audio message sent to the server.")
        except Exception as e:
            print(f"Failed to send audio message: {e}")

    async def truncate_item(self, item_id: str):
        message = {
            "action": "truncate_item",
            "id": item_id
        }
        await self.websocket.send(json.dumps(message))
        print(f"Truncated item {item_id}")

    async def delete_item(self, item_id: str):
        message = {
            "action": "delete_item",
            "id": item_id
        }
        await self.websocket.send(json.dumps(message))
        print(f"Deleted item {item_id}")

    async def receive(self):
        try:
            async for message in self.websocket:
                await self.handle_response(message)
        except websockets.exceptions.ConnectionClosed:
            print("Connection closed by server.")
        except Exception as e:
            print(f"Error occurred: {e}")

    async def handle_response(self, message: str):
        try:
            data = json.loads(message)
            msg_type = data.get("type")
            if msg_type == "session_created":
                self.session_id = data.get("session_id")
                print("Session created successfully.")
            elif msg_type == "text":
                text = data.get("content")
                print(f"Assistant: {text}")
            elif msg_type == "audio":
                encoded_audio = data.get("content")
                audio_bytes = base64.b64decode(encoded_audio)
                threading.Thread(target=play_audio, args=(audio_bytes,)).start()
            elif msg_type == "function_call":
                function_name = data.get("function_name")
                parameters = data.get("parameters")
                print(f"Calling function {function_name} with parameters: {parameters}")
            elif msg_type == "error":
                error_message = data.get("message")
                print(f"Error: {error_message}")
            else:
                print(f"Unknown message type: {msg_type}")
        except json.JSONDecodeError:
            print("Received non-JSON message.")
        except Exception as e:
            print(f"Failed to handle response: {e}")

    async def close_connection(self):
        self.running = False
        if self.websocket:
            await self.websocket.close()
            print("Connection closed.")

def main():
    client = Client()
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(client.connect())
    except KeyboardInterrupt:
        print("Interrupted by user.")
        loop.run_until_complete(client.close_connection())
    finally:
        p.terminate()
        loop.close()

if __name__ == "__main__":
    print("Starting client. Type '/exit' to quit.")
    print("Commands:")
    print("  /exit - Close the connection and exit.")
    print("  /truncate <item_id> - Truncate a conversation item.")
    print("  /delete <item_id> - Delete a conversation item.")
    print("  /audio <file_path> - Send an audio file as input.")
    main()

File: realtime/realtime.cs
using System;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NAudio.Wave;
using System.IO;
using System.Collections.Concurrent;

namespace OpenAIWebSocketClient
{
    // Define the structure of messages
    public class SessionCreateMessage
    {
        public string type = "session_create";
        public string model { get; set; }
        public string[] modalities { get; set; }
        public string voice { get; set; }
        public InputAudioConfig input_audio_config { get; set; }
        public OutputAudioConfig output_audio_config { get; set; }
        public TurnDetectionConfig turn_detection { get; set; }
    }

    public class InputAudioConfig
    {
        public string transcription_model { get; set; }
        public string format { get; set; }
    }

    public class OutputAudioConfig
    {
        public string format { get; set; }
    }

    public class TurnDetectionConfig
    {
        public string type { get; set; }
        public double threshold { get; set; }
        public int silence_duration_ms { get; set; }
    }

    public class ConversationItem
    {
        public string type { get; set; } = "user";
        public string content { get; set; }
    }

    public class TruncateDeleteEvent
    {
        public string type { get; set; }
        public string event_type { get; set; }
        public string item_id { get; set; }
    }

    class Program
    {
        private static ClientWebSocket _webSocket = new ClientWebSocket();
        private static CancellationTokenSource _cancellation = new CancellationTokenSource();
        private static string _sessionId;
        private static ConcurrentDictionary<string, string> _conversationItems = new ConcurrentDictionary<string, string>();
        private static MemoryStream _audioBuffer = new MemoryStream();

        static async Task Main(string[] args)
        {
            Console.WriteLine("Starting OpenAI WebSocket Client...");

            try
            {
                await ConnectWebSocket();

                // Start receiving messages
                var receiveTask = ReceiveMessages();

                // Start handling user input
                await HandleUserInput();

                // Wait for receive task to complete
                await receiveTask;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error occurred: {ex.Message}");
            }
            finally
            {
                await CloseWebSocket();
            }
        }

        private static async Task ConnectWebSocket()
        {
            var uri = new Uri("wss://api.openai.com/v1/assistants-streaming");
            await _webSocket.ConnectAsync(uri, CancellationToken.None);
            Console.WriteLine("WebSocket connected.");

            // Create session
            var sessionMessage = new SessionCreateMessage
            {
                model = "gpt-4o-realtime-preview-2024-10-01",
                modalities = new string[] { "text", "audio" },
                voice = "alloy",
                input_audio_config = new InputAudioConfig
                {
                    transcription_model = "whisper-1",
                    format = "pcm16"
                },
                output_audio_config = new OutputAudioConfig
                {
                    format = "pcm16"
                },
                turn_detection = new TurnDetectionConfig
                {
                    type = "server_vad",
                    threshold = 0.5,
                    silence_duration_ms = 200
                }
            };

            string sessionJson = JsonConvert.SerializeObject(sessionMessage);
            await SendMessage(sessionJson);
            Console.WriteLine("Session created successfully.");
        }

        private static async Task SendMessage(string message)
        {
            var bytes = Encoding.UTF8.GetBytes(message);
            var buffer = new ArraySegment<byte>(bytes);
            await _webSocket.SendAsync(buffer, WebSocketMessageType.Text, true, CancellationToken.None);
        }

        private static async Task ReceiveMessages()
        {
            var buffer = new byte[8192];

            try
            {
                while (_webSocket.State == WebSocketState.Open)
                {
                    var result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), _cancellation.Token);

                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        Console.WriteLine("Server initiated close. Closing connection.");
                        await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None);
                        break;
                    }

                    int count = result.Count;
                    using (var ms = new MemoryStream())
                    {
                        ms.Write(buffer, 0, count);
                        while (!result.EndOfMessage)
                        {
                            result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), _cancellation.Token);
                            ms.Write(buffer, 0, result.Count);
                        }
                        var message = Encoding.UTF8.GetString(ms.ToArray());
                        HandleServerMessage(message);
                    }
                }
            }
            catch (OperationCanceledException)
            {
                // Ignore cancellation
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error occurred: {ex.Message}");
            }
        }

        private static void HandleServerMessage(string message)
        {
            try
            {
                var json = JObject.Parse(message);
                var type = json["type"]?.ToString();

                switch (type)
                {
                    case "text":
                        string responseText = json["content"]?.ToString();
                        Console.WriteLine($"Assistant: {responseText}");
                        break;
                    case "audio":
                        string audioBase64 = json["content"]?.ToString();
                        PlayAudio(audioBase64);
                        break;
                    case "function_call":
                        string functionName = json["function_name"]?.ToString();
                        string parameters = json["parameters"]?.ToString();
                        Console.WriteLine($"Calling function {functionName} with parameters: {parameters}");
                        break;
                    case "error":
                        string errorMessage = json["message"]?.ToString();
                        Console.WriteLine($"Error: {errorMessage}");
                        break;
                    default:
                        Console.WriteLine($"Unknown message type: {type}");
                        break;
                }
            }
            catch (JsonException ex)
            {
                Console.WriteLine($"Error parsing server message: {ex.Message}");
            }
        }

        private static void PlayAudio(string audioBase64)
        {
            try
            {
                byte[] audioBytes = Convert.FromBase64String(audioBase64);
                using (var ms = new MemoryStream(audioBytes))
                using (var waveOut = new WaveOutEvent())
                using (var waveProvider = new RawSourceWaveStream(ms, new WaveFormat(16000, 16, 1)))
                {
                    waveOut.Init(waveProvider);
                    waveOut.Play();
                    while (waveOut.PlaybackState == PlaybackState.Playing)
                    {
                        Thread.Sleep(100);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error playing audio: {ex.Message}");
            }
        }

        private static async Task HandleUserInput()
        {
            Console.WriteLine("You can start typing your messages. Type 'exit' to quit.");
            Console.WriteLine("To send audio, type 'audio'.");

            while (true)
            {
                Console.Write("> ");
                string input = Console.ReadLine();

                if (string.IsNullOrWhiteSpace(input))
                    continue;

                if (input.Equals("exit", StringComparison.OrdinalIgnoreCase))
                {
                    break;
                }
                else if (input.StartsWith("truncate ", StringComparison.OrdinalIgnoreCase))
                {
                    string itemId = input.Substring(9).Trim();
                    await TruncateConversationItem(itemId);
                }
                else if (input.StartsWith("delete ", StringComparison.OrdinalIgnoreCase))
                {
                    string itemId = input.Substring(7).Trim();
                    await DeleteConversationItem(itemId);
                }
                else if (input.Equals("audio", StringComparison.OrdinalIgnoreCase))
                {
                    await HandleAudioInput();
                }
                else
                {
                    await SendTextMessage(input);
                }
            }
        }

        private static async Task SendTextMessage(string text)
        {
            var conversationItem = new ConversationItem
            {
                content = text
            };

            string json = JsonConvert.SerializeObject(conversationItem);
            await SendMessage(json);
            Console.WriteLine($"User said: {text}");
        }

        private static async Task HandleAudioInput()
        {
            Console.WriteLine("Recording audio... Press Enter to stop.");
            var audioBytes = RecordAudio();
            if (audioBytes != null && audioBytes.Length > 0)
            {
                string audioBase64 = Convert.ToBase64String(audioBytes);
                // Assuming the server expects audio messages with type 'audio'
                var audioMessage = new
                {
                    type = "audio",
                    content = audioBase64
                };
                string json = JsonConvert.SerializeObject(audioMessage);
                await SendMessage(json);
                Console.WriteLine("Audio message sent to the server.");
            }
            else
            {
                Console.WriteLine("No audio recorded.");
            }
        }

        private static byte[] RecordAudio()
        {
            try
            {
                using (var waveIn = new WaveInEvent())
                {
                    waveIn.WaveFormat = new WaveFormat(16000, 16, 1); // PCM16
                    using (var ms = new MemoryStream())
                    {
                        waveIn.DataAvailable += (s, a) =>
                        {
                            ms.Write(a.Buffer, 0, a.BytesRecorded);
                        };

                        waveIn.StartRecording();
                        Console.ReadLine(); // Wait for Enter to stop
                        waveIn.StopRecording();

                        return ms.ToArray();
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error recording audio: {ex.Message}");
                return null;
            }
        }

        private static async Task TruncateConversationItem(string itemId)
        {
            var truncateEvent = new TruncateDeleteEvent
            {
                type = "conversation_event",
                event_type = "truncate",
                item_id = itemId
            };

            string json = JsonConvert.SerializeObject(truncateEvent);
            await SendMessage(json);
            Console.WriteLine($"Truncated item {itemId}");
        }

        private static async Task DeleteConversationItem(string itemId)
        {
            var deleteEvent = new TruncateDeleteEvent
            {
                type = "conversation_event",
                event_type = "delete",
                item_id = itemId
            };

            string json = JsonConvert.SerializeObject(deleteEvent);
            await SendMessage(json);
            Console.WriteLine($"Deleted item {itemId}");
        }

        private static async Task CloseWebSocket()
        {
            if (_webSocket.State == WebSocketState.Open || _webSocket.State == WebSocketState.CloseReceived)
            {
                await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None);
                Console.WriteLine("Connection closed.");
            }
            _cancellation.Cancel();
            _webSocket.Dispose();
        }
    }
}
